docs/docs/modules/metadata.mdx

```diff
@@ -0,0 +1,400 @@
+
+## Direct use
+
+Upsides:
+* Flexible
+* Readable by other methods
+
+Downsides:
+* Can be quite tedious to write
+* Structural validation must be done by hand (partially)
+
+As the metadatum structure is fairly expressive, we can directly use it using the structs in the metadata module of this library. These directly represent the types given in the CDDL. Namely:
+* TransactionMetadatum - Can represent one of those 5 variant types.
+* MetadataMap - The map variant that maps from metadatums to other metadatums. This is unordered and indexed by metadatums. This is like an object in JSON.
+* MetadataList - An ordered list indexed starting at 0. This is like an array in JSON.
+
+The variants for numbers, bytes and text are not specific to metadata and are directly used with the general `Int` type representing a signed or unsigned number, byte arrays accepting byte arrays/`Buffer`, and strings being JS strings.
+
+We could construct the JSON example above with the following code:
+```javascript
+const map = CardanoWasm.MetadataMap.new();
+map.insert(
+  CardanoWasm.TransactionMetadatum.new_text("receiver_id"),
+  CardanoWasm.TransactionMetadatum.new_text("SJKdj34k3jjKFDKfjFUDfdjkfd"),
+);
+map.insert(
+  CardanoWasm.TransactionMetadatum.new_text("sender_id"),
+  CardanoWasm.TransactionMetadatum.new_text("jkfdsufjdk34h3Sdfjdhfduf873"),
+);
+map.insert(
+  CardanoWasm.TransactionMetadatum.new_text("comment"),
+  CardanoWasm.TransactionMetadatum.new_text("happy birthday"),
+);
+const tags = CardanoWasm.MetadataList.new();
+tags.add(CardanoWasm.TransactionMetadatum.new_int(CardanoWasm.Int.new(CardanoWasm.BigNum.from_str("0"))));
+tags.add(CardanoWasm.TransactionMetadatum.new_int(CardanoWasm.Int.new(CardanoWasm.BigNum.from_str("264"))));
+tags.add(CardanoWasm.TransactionMetadatum.new_int(CardanoWasm.Int.new_negative(CardanoWasm.BigNum.from_str("1024"))));
+tags.add(CardanoWasm.TransactionMetadatum.new_int(CardanoWasm.Int.new(CardanoWasm.BigNum.from_str("32"))));
+map.insert(
+  CardanoWasm.TransactionMetadatum.new_text("tags"),
+  CardanoWasm.TransactionMetadatum.new_list(tags),
+);
+const metadatum = CardanoWasm.TransactionMetadatum.new_map(map);
+```
+
+We could then parse the information back as such:
+```javascript
+try {
+  const map = metadatum.as_map();
+  const receiver = map.get(CardanoWasm.TransactionMetadatum.new_text("receiver_id"));
+  const sender = map.get(CardanoWasm.TransactionMetadatum.new_text("sender_id"));
+  const comment = map.get(CardanoWasm.TransactionMetadatum.new_text("comment"));
+  const tags = map.get(CardanoWasm.TransactionMetadatum.new_text("tags"));
+} catch (e) {
+  // structure did not match
+}
+```
+
+For decoding in a more exploratory manner we can check the types first as such:
+```javascript
+function parseMetadata(metadata) {
+  // we must check the type first to know how to handle it
+  switch (metadata.kind()) {
+    case CardanoWasm.TransactionMetadatumKind.MetadataMap:
```

Comment:
this code will need to be updated. the variants here are called `Map` and `List` now, not `MetadataMap`.